### ISSUE TITLE
Traffic-provBZ fails with deserialization error

### ISSUE THREAD

**Issue by clezag** [noi-techpark/bdp-commons#682]
Traffic provbz datacollector fails with some kind of deserialization error

`
{"@timestamp":"2024-12-18T14:50:27.445Z","message":"Unexpected error occurred in scheduled task","level":"ERROR","level_value":40000,"stack_trace":"com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `[Lit.bz.odh.trafficprovbz.dto.AggregatedDataDto;` from Object value (token `JsonToken.START_OBJECT`)\n at [Source: (String)\"{\"error\":{\"message\":\"expected:'Number Token', actual:':', at offset:1465\",\"exception\":\"UnexpectedElasticsearchClientException\"}}\"; line: 1, column: 1]\n\tat com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)\n\tat com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1741)\n\tat com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1515)\n\tat com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1462)\n\tat com.fasterxml.jackson.databind.deser.std.ObjectArrayDeserializer.handleNonArray(ObjectArrayDeserializer.java:343)\n\tat com.fasterxml.jackson.databind.deser.std.ObjectArrayDeserializer.deserialize(ObjectArrayDeserializer.java:194)\n\tat com.fasterxml.jackson.databind.deser.std.ObjectArrayDeserializer.deserialize(ObjectArrayDeserializer.java:24)\n\tat com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)\n\tat com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4674)\n\tat com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3629)\n\tat com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3597)\n\tat it.bz.odh.trafficprovbz.FamasClient.getAggregatedDataOnStations(FamasClient.java:129)\n\tat it.bz.odh.trafficprovbz.FamasClient$$FastClassBySpringCGLIB$$5b77abb1.invoke(<generated>)\n\tat org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)\n\tat org.springframework.aop.framework.CglibAopProxy.invokeMethod(CglibAopProxy.java:386)\n\tat org.springframework.aop.framework.CglibAopProxy.access$000(CglibAopProxy.java:85)\n\tat org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:704)\n\tat it.bz.odh.trafficprovbz.FamasClient$$EnhancerBySpringCGLIB$$ec030f64.getAggregatedDataOnStations(<generated>)\n\tat it.bz.odh.trafficprovbz.SyncScheduler.syncJobTrafficMeasurements(SyncScheduler.java:235)\n\tat it.bz.odh.trafficprovbz.SyncScheduler.sync(SyncScheduler.java:162)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:84)\n\tat org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)\n\tat org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)\n\tat java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n","application":"DCTEST","application_version":"0000"}
`

**Comment by clezag** [comment]
I've added some additional logging, and it looks like the remote API is throwing an error and thus returning an unexpected JSON:

<code>curl -X POST https://webservices.trafficopab.famassystem.it/api/v1/DatiAggregatiSuPostazioni -u '*****:*****' --data '{"IdPostazioni":["3"],"InizioPeriodo":"2024-12-18T04:20:15Z","FinePeriodo":"2024-12-19T11:50:58Z"}' -H 'Content-Type: application/json'
</code>

returns
```json
{"error":{"message":"expected:'Number Token', actual:':', at offset:1383","exception":"UnexpectedElasticsearchClientException"}}
```

@rcavaliere I think we need to contact the data provider here

**Comment by rcavaliere** [comment]
@clezag thanks, I imagined here an issue of the Data Provider. I will contact them and let you know.

**Comment by clezag** [comment]
After they fixed the original server error, the next issue was that all the JSON fields went from being uppercase to lowercase.
I've changed all the field names in the data collector to lowercase, which made it work again (in testing).

There was also another rather ugly change: The date format randomly changes between `2025-01-27T09:40:00` and `2025-01-27T09:40:00Z` **within the same JSON response**. This is dangerous, because the second one (correctly) specifies the timezone as UTC, but the first one does not have any timezone information attached at all.

As a workaround, we now ignore the 'Z' and assume all are in UTC, but this is something we probably should report to them as a bug.

example from testing (note `data` field):
```json
   {
        "idPostazione": 3,
        "data": "2025-01-27T09:40:00",
        "corsia": 1,
        "direzione": "discendente",
        "totaleVeicoli": 55,
        "totaliPerClasseVeicolare": {
            "2": 41,
            "4": 8,
            "8": 2,
            "1": 1,
            "6": 1,
            "7": 1,
            "9": 1
        },
        "mediaArmonicaVelocita": 68.2,
        "headwayMedioSecondi": 4.49,
        "varianzaHeadwayMedioSecondi": 33.21,
        "gapMedioSecondi": 4.13,
        "varianzaGapMedioSecondi": 33.54
    },
    {
        "idPostazione": 3,
        "data": "2025-01-27T09:45:00",
        "corsia": 0,
        "direzione": "ascendente",
        "totaleVeicoli": 54,
        "totaliPerClasseVeicolare": {
            "2": 45,
            "4": 3,
            "1": 2,
            "6": 2,
            "5": 1,
            "8": 1
        },
        "mediaArmonicaVelocita": 72.2,
        "headwayMedioSecondi": 5.97,
        "varianzaHeadwayMedioSecondi": 69.67,
        "gapMedioSecondi": 5.67,
        "varianzaGapMedioSecondi": 70.05
    },
    {
        "idPostazione": 3,
        "data": "2025-01-27T09:45:00Z",
        "corsia": 0,
        "direzione": "discendente",
        "totaleVeicoli": 0
    },
    {
        "idPostazione": 3,
        "data": "2025-01-27T09:45:00Z",
        "corsia": 1,
        "direzione": "ascendente",
        "totaleVeicoli": 0
    },
    {
        "idPostazione": 3,
        "data": "2025-01-27T09:45:00",
        "corsia": 1,
        "direzione": "discendente",
        "totaleVeicoli": 45,
        "totaliPerClasseVeicolare": {
            "2": 39,
            "8": 3,
            "4": 1,
            "5": 1,
            "7": 1
        },
        "mediaArmonicaVelocita": 56.5,
        "headwayMedioSecondi": 14.64,
        "varianzaHeadwayMedioSecondi": 3136.94,
        "gapMedioSecondi": 14.25,
        "varianzaGapMedioSecondi": 3139.46
    },
```

**Comment by rcavaliere** [comment]
@clezag thanks for the fix, seems ok in testing, in my opinion you can bring this in production.  For the timestamp bug, I will provide the feedback to the technical supplier, in the meantime I agree with your implemented solution

**Comment by rcavaliere** [comment]
@clezag I see the data, good! But we have a big hole in the time series. I don't remember exactly, don't we have from the API the possibility to get the historical data?

**Comment by clezag** [comment]
@rcavaliere Yes, the code has some kind of history import mode, I will look into it.

**Comment by clezag** [comment]
@rcavaliere I've completed the import in testing, it's now running in production (will take a few hours).

**Comment by rcavaliere** [comment]
@clezag wonderful, I see the full timeseries in testing, thanks for this
